File: go/oauth.go
package threads

import (
	"context"
	"errors"
	"net/url"
	"strings"
)

// OAuth grant-type constants used by the Threads token endpoints.
const (
	grantAuthorizationCode = "authorization_code"
	grantExchangeToken     = "th_exchange_token"
	grantRefreshToken      = "th_refresh_token"
)

// DefaultAuthorizeBaseURL is the host that serves the OAuth consent UI.
// The token-exchange + refresh endpoints live on the Graph host (BaseURL).
const DefaultAuthorizeBaseURL = "https://threads.net"

// Scope is a Threads OAuth permission. Request the minimal set required.
type Scope string

// Recognised Threads OAuth scopes.
const (
	ScopeBasic          Scope = "threads_basic"
	ScopeContentPublish Scope = "threads_content_publish"
	ScopeManageInsights Scope = "threads_manage_insights"
	ScopeReadReplies    Scope = "threads_read_replies"
	ScopeManageReplies  Scope = "threads_manage_replies"
	ScopeKeywordSearch  Scope = "threads_keyword_search"
	ScopeManageMentions Scope = "threads_manage_mentions"
	ScopeDelete         Scope = "threads_delete"
)

// AuthorizeParams configures AuthorizationURL.
type AuthorizeParams struct {
	ClientID    string  // Threads app client id (required).
	RedirectURI string  // Registered redirect URI (required).
	Scopes      []Scope // At least one scope (required).
	State       string  // Optional CSRF / round-trip token.
	// AuthorizeBaseURL overrides the consent-UI host. Defaults to
	// DefaultAuthorizeBaseURL when empty.
	AuthorizeBaseURL string
}

// AuthorizationURL builds the URL to redirect the user to for consent. It
// performs no I/O. The user is sent here; on approval Threads redirects back
// to RedirectURI with a one-time ?code= the caller passes to ExchangeCode.
func AuthorizationURL(p AuthorizeParams) (string, error) {
	if p.ClientID == "" {
		return "", errors.New("threads: AuthorizationURL requires ClientID")
	}
	if p.RedirectURI == "" {
		return "", errors.New("threads: AuthorizationURL requires RedirectURI")
	}
	if len(p.Scopes) == 0 {
		return "", errors.New("threads: AuthorizationURL requires at least one Scope")
	}
	scopes := make([]string, len(p.Scopes))
	for i, s := range p.Scopes {
		scopes[i] = string(s)
	}
	base := strings.TrimRight(p.AuthorizeBaseURL, "/")
	if base == "" {
		base = DefaultAuthorizeBaseURL
	}
	q := url.Values{}
	q.Set("client_id", p.ClientID)
	q.Set("redirect_uri", p.RedirectURI)
	q.Set("scope", strings.Join(scopes, ","))
	q.Set("response_type", "code")
	if p.State != "" {
		q.Set("state", p.State)
	}
	return base + "/oauth/authorize?" + q.Encode(), nil
}

// ShortLivedToken is the result of exchanging an authorization code.
type ShortLivedToken struct {
	AccessToken string `json:"access_token"`
	UserID      int64  `json:"user_id,omitempty"`
}

// LongLivedToken is the result of exchanging a short-lived token, or of
// refreshing a long-lived token. Threads long-lived tokens last ~60 days.
type LongLivedToken struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type,omitempty"`
	ExpiresIn   int64  `json:"expires_in,omitempty"`
}

// ExchangeCodeParams carries the inputs to the code->token exchange.
type ExchangeCodeParams struct {
	ClientID     string
	ClientSecret string
	RedirectURI  string
	Code         string
}

// ExchangeCode swaps a one-time authorization code for a short-lived access
// token. This is the first server-side step of the OAuth flow and mirrors the
// backend's in-tree threads.Client.ExchangeCode. The client secret is sent in
// a form-encoded POST body over TLS and is never logged by this package.
func (c *Client) ExchangeCode(ctx context.Context, p ExchangeCodeParams) (*ShortLivedToken, error) {
	if p.ClientID == "" || p.ClientSecret == "" {
		return nil, errors.New("threads: ExchangeCode requires ClientID and ClientSecret")
	}
	if p.Code == "" {
		return nil, errors.New("threads: ExchangeCode requires Code")
	}
	if p.RedirectURI == "" {
		return nil, errors.New("threads: ExchangeCode requires RedirectURI")
	}
	form := url.Values{}
	form.Set("client_id", p.ClientID)
	form.Set("client_secret", p.ClientSecret)
	form.Set("grant_type", grantAuthorizationCode)
	form.Set("redirect_uri", p.RedirectURI)
	form.Set("code", p.Code)

	out := &ShortLivedToken{}
	if err := c.doOAuthPost(ctx, "/oauth/access_token", form, out); err != nil {
		return nil, err
	}
	return out, nil
}

// ExchangeForLongLivedParams carries the inputs to the short->long exchange.
type ExchangeForLongLivedParams struct {
	ClientSecret    string
	ShortLivedToken string
}

// ExchangeForLongLived swaps a short-lived token for a long-lived (~60 day)
// token via GET /access_token?grant_type=th_exchange_token.
func (c *Client) ExchangeForLongLived(ctx context.Context, p ExchangeForLongLivedParams) (*LongLivedToken, error) {
	if p.ClientSecret == "" {
		return nil, errors.New("threads: ExchangeForLongLived requires ClientSecret")
	}
	if p.ShortLivedToken == "" {
		return nil, errors.New("threads: ExchangeForLongLived requires ShortLivedToken")
	}
	q := url.Values{}
	q.Set("grant_type", grantExchangeToken)
	q.Set("client_secret", p.ClientSecret)
	q.Set("access_token", p.ShortLivedToken)

	out := &LongLivedToken{}
	if err := c.doOAuthGet(ctx, "/access_token", q, out); err != nil {
		return nil, err
	}
	return out, nil
}

// RefreshTokenParams carries the inputs to the Threads token refresh endpoint.
//
// Threads issues long-lived tokens that can be refreshed via
// GET /refresh_access_token while still valid. The endpoint requires only the
// access token itself (no client_secret).
type RefreshTokenParams struct {
	AccessToken string // The current long-lived access token to refresh.
}

// RefreshTokenResult is the decoded response. Retained as a distinct type for
// backwards compatibility; it is structurally identical to LongLivedToken.
type RefreshTokenResult struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	ExpiresIn   int64  `json:"expires_in"`
}

// RefreshAccessToken exchanges a still-valid long-lived token for a refreshed
// version. This call does not consume the *Client's stored AccessToken so the
// caller can refresh tokens on behalf of any user without reconfiguring.
func (c *Client) RefreshAccessToken(ctx context.Context, p RefreshTokenParams) (*RefreshTokenResult, error) {
	if p.AccessToken == "" {
		return nil, errors.New("threads: RefreshTokenParams requires AccessToken")
	}
	q := url.Values{}
	q.Set("grant_type", grantRefreshToken)
	q.Set("access_token", p.AccessToken)

	out := &RefreshTokenResult{}
	if err := c.doOAuthGet(ctx, "/refresh_access_token", q, out); err != nil {
		return nil, err
	}
	return out, nil
}

File: threads/oauth.py
"""Threads OAuth 2.0 client — authorization URL, code exchange, token refresh.

This is an *app-level* flow: it uses your Threads app's ``client_id`` and
``client_secret`` to mint and refresh user access tokens. It does not require a
pre-existing user token (that is what it produces), so it is implemented as a
standalone client rather than a namespace on :class:`~threads.client.ThreadsClient`.

Flow overview::

    1. authorization_url(...)          -> redirect the user, receive ?code=...
    2. exchange_code(code)             -> ShortLivedToken (~1 hour)
    3. exchange_for_long_lived(token)  -> LongLivedToken  (~60 days)
    4. refresh_long_lived(token)       -> LongLivedToken  (refreshed ~60 days)

No vendor secret is ever logged or embedded; credentials are supplied by the
caller (sourced from environment/secret store) and only ever sent over TLS to
the Threads OAuth host.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import urlencode

import httpx

from threads.config import (
    DEFAULT_MAX_CONNECTIONS,
    DEFAULT_MAX_KEEPALIVE_CONNECTIONS,
    DEFAULT_TIMEOUT,
    THREADS_AUTHORIZE_BASE_URL,
    THREADS_OAUTH_BASE_URL,
)
from threads.exceptions import (
    ThreadsAPIError,
    ThreadsConfigError,
    build_api_error,
)
from threads.models.oauth import LongLivedToken, ShortLivedToken, ThreadsScope

_GRANT_AUTHORIZATION_CODE = "authorization_code"
_GRANT_EXCHANGE_TOKEN = "th_exchange_token"
_GRANT_REFRESH_TOKEN = "th_refresh_token"


class ThreadsOAuth:
    """Client for the Threads OAuth 2.0 authorization-code flow.

    Parameters
    ----------
    client_id:
        Your Threads app's client id (Meta App ID).
    client_secret:
        Your Threads app's client secret. Sourced from a secret store /
        environment variable — never hard-code it.
    redirect_uri:
        The redirect URI registered with your Threads app.
    timeout:
        Total per-request timeout in seconds.
    authorize_base_url / oauth_base_url:
        Overridable hosts (used by tests to point at a mock transport).
    transport:
        Optional :class:`httpx.AsyncBaseTransport` injected for testing.
    """

    __slots__ = (
        "_client_id",
        "_client_secret",
        "_redirect_uri",
        "_authorize_base_url",
        "_oauth_base_url",
        "_timeout",
        "_transport",
        "_client",
    )

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        redirect_uri: str,
        *,
        timeout: float = DEFAULT_TIMEOUT,
        authorize_base_url: str = THREADS_AUTHORIZE_BASE_URL,
        oauth_base_url: str = THREADS_OAUTH_BASE_URL,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        if not client_id:
            raise ThreadsConfigError("client_id must not be empty.")
        if not client_secret:
            raise ThreadsConfigError("client_secret must not be empty.")
        if not redirect_uri:
            raise ThreadsConfigError("redirect_uri must not be empty.")

        self._client_id = client_id
        self._client_secret = client_secret
        self._redirect_uri = redirect_uri
        self._authorize_base_url = authorize_base_url.rstrip("/")
        self._oauth_base_url = oauth_base_url.rstrip("/")
        self._timeout = httpx.Timeout(timeout)
        self._transport = transport
        self._client = httpx.AsyncClient(
            base_url=self._oauth_base_url,
            timeout=self._timeout,
            limits=httpx.Limits(
                max_connections=DEFAULT_MAX_CONNECTIONS,
                max_keepalive_connections=DEFAULT_MAX_KEEPALIVE_CONNECTIONS,
            ),
            transport=self._transport,
        )

    async def __aenter__(self) -> ThreadsOAuth:
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        """Close the underlying HTTP client and release pooled connections."""
        if not self._client.is_closed:
            await self._client.aclose()

    # -- step 1: authorization URL -------------------------------------------

    def authorization_url(
        self,
        scopes: list[ThreadsScope],
        *,
        state: str | None = None,
    ) -> str:
        """Build the URL to redirect the user to for consent.

        Parameters
        ----------
        scopes:
            One or more :class:`~threads.models.oauth.ThreadsScope` values.
        state:
            Opaque CSRF / round-trip token echoed back to your redirect URI.
        """
        if not scopes:
            raise ThreadsConfigError("authorization_url requires at least one scope.")
        params: dict[str, str] = {
            "client_id": self._client_id,
            "redirect_uri": self._redirect_uri,
            "scope": ",".join(scope.value for scope in scopes),
            "response_type": "code",
        }
        if state is not None:
            params["state"] = state
        return f"{self._authorize_base_url}/oauth/authorize?{urlencode(params)}"

    # -- step 2: code -> short-lived token ------------------------------------

    async def exchange_code(self, code: str) -> ShortLivedToken:
        """Exchange an authorization ``code`` for a short-lived access token."""
        if not code:
            raise ThreadsConfigError("exchange_code requires a non-empty code.")
        data = {
            "client_id": self._client_id,
            "client_secret": self._client_secret,
            "grant_type": _GRANT_AUTHORIZATION_CODE,
            "redirect_uri": self._redirect_uri,
            "code": code,
        }
        body = await self._post("/oauth/access_token", data)
        return ShortLivedToken.model_validate(body)

    # -- step 3: short-lived -> long-lived token ------------------------------

    async def exchange_for_long_lived(self, short_lived_token: str) -> LongLivedToken:
        """Exchange a short-lived token for a long-lived (~60 day) token."""
        if not short_lived_token:
            raise ThreadsConfigError(
                "exchange_for_long_lived requires a non-empty short_lived_token."
            )
        params = {
            "grant_type": _GRANT_EXCHANGE_TOKEN,
            "client_secret": self._client_secret,
            "access_token": short_lived_token,
        }
        body = await self._get("/access_token", params)
        return LongLivedToken.model_validate(body)

    # -- step 4: refresh a long-lived token -----------------------------------

    async def refresh_long_lived(self, long_lived_token: str) -> LongLivedToken:
        """Refresh a still-valid long-lived token for a new ~60 day token.

        Threads does not issue a separate refresh token; the existing
        long-lived access token is presented to ``/refresh_access_token``.
        """
        if not long_lived_token:
            raise ThreadsConfigError("refresh_long_lived requires a non-empty long_lived_token.")
        params = {
            "grant_type": _GRANT_REFRESH_TOKEN,
            "access_token": long_lived_token,
        }
        body = await self._get("/refresh_access_token", params)
        return LongLivedToken.model_validate(body)

    # -- internal request helpers --------------------------------------------

    async def _get(self, path: str, params: dict[str, str]) -> dict[str, Any]:
        resp = await self._client.get(path, params=params)
        return self._parse_response(resp)

    async def _post(self, path: str, data: dict[str, str]) -> dict[str, Any]:
        resp = await self._client.post(path, data=data)
        return self._parse_response(resp)

    @staticmethod
    def _parse_response(resp: httpx.Response) -> dict[str, Any]:
        try:
            body: dict[str, Any] = resp.json()
        except ValueError as exc:
            raise ThreadsAPIError(
                f"HTTP {resp.status_code}",
                http_status=resp.status_code,
            ) from exc

        if isinstance(body, dict) and "error" in body:
            err = body["error"]
            # Threads OAuth errors can be either the Graph nested envelope
            # ({"error": {"message": ..., "code": ...}}) or the flat OAuth2
            # form ({"error": "invalid_grant", "error_description": ...}).
            if isinstance(err, dict):
                raise build_api_error(
                    message=err.get("message", "Unknown error"),
                    http_status=resp.status_code,
                    error_type=err.get("type", ""),
                    error_code=err.get("code", 0),
                    error_subcode=err.get("error_subcode", 0),
                    fbtrace_id=err.get("fbtrace_id", ""),
                )
            raise build_api_error(
                message=str(body.get("error_description", err)),
                http_status=resp.status_code,
                error_type=str(err),
                error_code=0,
                error_subcode=0,
                fbtrace_id="",
            )

        if resp.status_code >= 400:
            raise ThreadsAPIError(
                f"HTTP {resp.status_code}",
                http_status=resp.status_code,
            )

        return body
